module/zfs/zpl_inode.c

```diff
@@ -708,6 +708,44 @@ zpl_revalidate(struct dentry *dentry, unsigned int flags)
 	return (1);
 }
 
+/*
+ * Valid FIEMAP flags:
+ * - FIEMAP_FLAG_SYNC    - Sync extents before reporting
+ * - FIEMAP_FLAG_XATTR   - Report extents used by xattrs (Unsupported)
+ * - FIEMAP_FLAG_COPIES  - Report all data copies (ZFS-only)
+ * - FIEMAP_FLAG_NOMERGE - Never merge blocks in to extents (ZFS-only)
+ */
+static int
+zpl_fiemap(struct inode *ip, struct fiemap_extent_info *fei,
+    __u64 start, __u64 len)
+{
+	zfs_fiemap_t *fm;
+	unsigned int flags = fei->fi_flags;
+	fstrans_cookie_t cookie;
+	int error = 0;
+
+	/* Incompatible ZFS-only flags masked out of compatibility check */
+	fei->fi_flags &= ~ZFS_FIEMAP_FLAGS_ZFS;
+
+	error = fiemap_check_flags(fei, ZFS_FIEMAP_FLAGS_COMPAT);
```

Comment:
You don't have to call `fiemap_check_flags()` if you don't want.  Make a `zpl_check_flags()` function that checks for all of the flags that the code currently understands.  I made sure that this function was not called at the VFS layer, exactly so that we could have different flags on a per-filesystem basis.